sdk/src/types/block/slot/epoch.rs

```diff
@@ -32,52 +31,47 @@ use crate::types::block::Error;
 /// | 2  | 16  | 24  |
 // ...
 #[derive(
-    Copy, Clone, Debug, Default, Eq, PartialEq, Ord, PartialOrd, Hash, From, Deref, Display, FromStr, packable::Packable,
+    Copy,
+    Clone,
+    Debug,
+    Default,
+    Eq,
+    PartialEq,
+    Ord,
+    PartialOrd,
+    Hash,
+    From,
+    Deref,
+    Add,
+    AddAssign,
+    Sub,
+    SubAssign,
+    Display,
+    FromStr,
+    packable::Packable,
 )]
 #[repr(transparent)]
-pub struct EpochIndex(u64);
+pub struct EpochIndex(pub u64);
 
 impl EpochIndex {
-    /// Creates a new [`EpochIndex`].
-    pub fn new(index: u64) -> Self {
-        Self::from(index)
+    /// Gets the range of slots this epoch contains.
+    pub fn slot_index_range(&self, slots_per_epoch_exponent: u8) -> core::ops::RangeInclusive<SlotIndex> {
+        self.first_slot_index(slots_per_epoch_exponent)..=self.last_slot_index(slots_per_epoch_exponent)
     }
 
     /// Gets the epoch index given a [`SlotIndex`].
-    pub fn from_slot_index(
-        slot_index: SlotIndex,
-        slots_per_epoch_exponent_iter: impl Iterator<Item = (Self, u8)>,
-    ) -> Result<Self, Error> {
-        let mut slot_index = *slot_index;
-        let mut res = 0;
-        let mut last = None;
-        for (start_epoch, exponent) in slots_per_epoch_exponent_iter {
-            if let Some((last_start_epoch, last_exponent)) = last {
-                if *start_epoch <= last_start_epoch {
-                    return Err(Error::InvalidStartEpoch(start_epoch));
-                }
-                // Get the number of slots this range of epochs represents
-                let slots_in_range = (*start_epoch - last_start_epoch) << last_exponent;
-                // Check whether the slot index is contained in this range
-                if slot_index > slots_in_range {
-                    // Update the slot index so it is in the context of the next epoch
-                    slot_index -= slots_in_range;
-                } else {
-                    break;
-                }
-            }
-            if *start_epoch > res {
-                // We can't calculate the epoch if we don't have the exponent for the containing range
-                if slot_index > 0 {
-                    return Err(Error::InvalidStartEpoch(start_epoch));
-                } else {
-                    break;
-                }
-            }
-            res = *start_epoch + (slot_index >> exponent);
-            last = Some((*start_epoch, exponent));
-        }
-        Ok(Self(res))
+    pub fn from_slot_index(slot_index: SlotIndex, slots_per_epoch_exponent: u8) -> Self {
+        Self(*slot_index >> slots_per_epoch_exponent)
+    }
```

Comment:
I know you didn't write the doc, but consider fixing it since it's not a getter.

Reply:
I think "Creates the epoch index given ..." would be SLIGHTLY better since it creates a `Self` and `Gets` usually indicates a getter, but I don't care that much. Just a small detail I couldn't unsee anymore :smile: